### PR TITLE
Low-power wakeup feature & GPIO refactoring

### PR DESCRIPTION
**Pull request by dotcypress**
This PR contains an initial implementation of wakeup pins/lanes and GPIO pins refactoring.

### CODE REVIEW

File: src/power.rs [hunk 1]
```diff
@@ -56,6 +138,27 @@ impl Power {
     }
 }
 
+macro_rules! wakeup_pins {
+    ($($PIN:path: $line:expr,)+) => {
+        $(
+            impl<M> Into<WakeUpLine> for &$PIN {
```
Comment (andresv) [inline src/power.rs]: Hmm clippy thinks that it is better to use `From`.

Comment (dotcypress) [reply src/power.rs]: Yup, fixed!

File: src/power.rs [hunk 2]
```diff
@@ -18,6 +19,17 @@ pub enum PowerMode {
...
 
+pub enum WakeUpLine {
+    Internal,
+    WKUP1,
+    WKUP2,
+    // TODO: missing in SVD file
+    // WKUP3,
+    WKUP4,
+    WKUP5,
+    WKUP6,
```
Comment (andresv) [inline src/power.rs:30]: Does it make sense to use rust idiomatic `Wkup1` or it probably makes it actually more confusing for the user.

File: src/power.rs [hunk 3]
```diff
@@ -28,6 +40,76 @@ impl Power {
...
+                self.rb.cr3.modify(|_, w| w.ewup4().set_bit());
+                self.rb.cr4.modify(|_, w| w.wp4().bit(edge));
+            }
+            WakeUpLine::WKUP5 => {
+                self.rb.cr3.modify(|_, w| w.ewup5().set_bit());
+                self.rb.cr4.modify(|_, w| w.wp4().bit(edge));
+            }
+            WakeUpLine::WKUP6 => {
+                self.rb.cr3.modify(|_, w| w.ewup6().set_bit());
+                self.rb.cr4.modify(|_, w| w.wp5().bit(edge));
```
Comment (andresv) [inline src/power.rs:96]: Did not check from datasheet but are `wp4` and `wp5` ok here (1, 2, 4 are different)?

File: src/power.rs (hunk 2 above)
Comment (dotcypress) [reply src/power.rs]: Good point!
I was using abbreviation from datasheet, but yes, you right, we must use more idiomatic and self-descriptive names.

File: src/power.rs (hunk 3 above)
Comment (dotcypress) [reply src/power.rs]: Oops, fixed

File: src/power.rs [hunk 4]
```diff
@@ -28,6 +39,76 @@ impl Power {
...
+            WakeUp::Line6 => self.rb.scr.write(|w| w.cwuf6().set_bit()),
+            _ => {}
+        }
+    }
+
+    pub fn clear_standby_flag(&mut self) {
+        self.rb.scr.write(|w| w.csbf().set_bit());
+    }
+
+    pub fn enable_wakeup_lane<L: Into<WakeUp>>(&mut self, lane: L, edge: SignalEdge) {
```
Comment (Disasm) [inline src/power.rs:72]: I guess all these functions consume a GPIO pin without a way to get it back?

File: src/power.rs [hunk 5]
```diff
@@ -56,6 +137,27 @@ impl Power {
...
+                fn from(_: &$PIN) -> Self {
+                    $line
+                 }
+            }
+        )+
+    }
+}
+
+wakeup_pins! {
+    gpioa::PA0<M>: WakeUp::Line1,
```
Comment (Disasm) [inline src/power.rs:153]: I'd move `<M>` to the macro above

Comment (dotcypress) [reply src/power.rs:153]: I can't do that, cause macro require full type :-(

File: src/power.rs (hunk 4 above)
Comment (dotcypress) [reply src/power.rs:72]: Nope, `From` trait implemented for `&$PIN`

File: src/power.rs (hunk 5 above)
Comment (burrbull) [reply src/power.rs:153]: Reexport all `PXX` types from `gpio::gpiox` to `gpio`. This simplifies a lot of things

Comment (burrbull) [reply src/power.rs:153]: https://github.com/stm32-rs/stm32f4xx-hal/blob/12b5bc13ab0b5ce96f01cb239e76aecf64d396d7/src/gpio.rs#L522

**Review by andresv — APPROVED** [review]

**Pull request merged**